### DISCUSSION TITLE
Suggestions fields meaning

### DISCUSSION THREAD

**Discussion by Hazoooka** [sn4k3/UVtools#733] [General]
No manual explaining the suggestions settings.
It's not possible to figure out what the settings influence.
There are the pop-up informations under the cursor (on hover) but not explaining anything.
Some settings are described as minimum and maximum but no idea what parameters to use and how will it influence the print settings. For example "trigger detection", no explanation what it is related to and what it triggers.

Please make a clear and detailed explanation of every parameter in the suggestions, every section, every setting.

### How to reproduce

Open the Suggestions settings and try to figure out what they mean with no knowledge of the maker's intention.

### Files

![Suggestions bottom layers](https://github.com/sn4k3/UVtools/assets/137073500/b8b65640-6ca2-4d55-9795-9684009e5939)
![Suggestions transition layers](https://github.com/sn4k3/UVtools/assets/137073500/df7f08d5-f4a2-4c9e-a4a3-b181524fa1dd)
![Suggestions wait before print](https://github.com/sn4k3/UVtools/assets/137073500/6a14e0c1-2f1f-405f-8f79-6ac695cc9c93)

**Comment by Hazoooka** [comment]
OK, so let's go over every setting.
Let's start from the beginning.

**Comment by sn4k3** [comment]
## Limits

The "Limit" field serves two purposes:

1. Constrains the value which the suggestion is trying to set, for example: Transition layer count, if your configuration runs off the set limits it will constrain to Min/Max. Ex: Limit from 3 to 10, your configuration leads to 2 resulting layers, then it will constrain to 3, if resulting in 20 layers it will constrain to 10.
2. The "trigger detection", valid only when you select "Outside limits" in "Apply when", that will test the file value which suggestion applies and if outside the limit it will trigger the suggestion. If the file value is inside the limit, it will pass the suggestion and classify as "applied", even if suggestion optimal value is different, as it is inside your defined limit, in other words, the range you consider safe to have. Example: Limit from 3 to 10 transition layers. Your file have 5 transition layers, then suggestion will not trigger nor allow to apply because is inside your desired range. All values: 3,4,5,6,7,8,9,10 will pass. Any value outside the limit will trigger the suggestion for user to apply.

## Proportional fields

When you see a proportional field, it uses the first-grade math technic where:

```
Type A - Type B
     X - Type B2

X = Type A * Type B2 / Type B
Where X is Type A2 or what you want to know
```

So on your screenshot example you have:
```
Proportional wait time: 10s
Proportional layer area: 10000mm2
```

It means that your optimal time is 10s for a 10000mm2 layer area, in other words, a layer area of 10000mm2 will have 10s set of wait time.
Then all diferent layer areas will corelate.
Let's say layer 10 have an area of 3000mm2, we need to know which wait time to set because you want time relative to mass, going back to math:

```
10s - 10000mm2
 X  - 3000mm2

X = 10 * 3000 / 10000 = 3s
```

Layer 10 will set 3s of wait time, but don't forget that defined limits always constrain values, so wait time will never run off your set limit range. So in the end Layer 10 Wait Time = 3s constrained in your limit range.

**Comment by Hazoooka** [comment]
Hi Tiago.
Thank you.

First thing: get someone to help you with designing the UVtools, seriously.
Second: do not write the manual, ever!
I haven't seen such an overcomplicated settings and explanation in my life!
You really need to consult what you're doing with someone, or with the community.

Now to the subject, the only thing I have understood is that I can get the program to set the values based on some other parameters.
But what if I want to go simple, like most of users do, and set my own parameters, and forget about what the program says.
Say, I want to set 30s wait time for the two base layers (easy to do by hand) and then set the normal layers to go from 30s to 1.5s through 30 transition layers. And I don't care about the area, just the straight forward simple transition of wait time.
How do I do this?

**Reply by sn4k3** [reply]
> I haven't seen such an overcomplicated settings and explanation in my life!

Is it? IMO they aren't so complicated and those were requested by people, in fact apply wait time per mass is the correct thing as study show that the mass affect the required wait time for EF stabilization but also optimize times, and for that you need this "complicated" stuff, or else you will never get it by hand. The difference is with UVtools is that you have to power over every setting, no hidden vars, no magic, good or bad you are in control of every set values. For those who do not want to tinker with settings, UVtools base settings for suggestions are good to hit "ok-and-forget", you don't even need to know what they are, it will boost print quality with a click of a button and that way you got the "easy-mode", but it looks likes none of that are good enough to you, and so I recommend you to stick to slicer only as you find everything in UVtools a problem and too much hard/a pain to deal with.

> And I don't care about the area, just the straight forward simple transition of wait time.

Then on "Set type" configure it to use "Fixed Time" and input the number of transition layer below. How hard is that?

**Comment by Hazoooka** [comment]
Yes, you need to work on being clear with what you are offering in the program.
Having control over the variety of areas in layers is logical, just the explanation is not understandable for everyone.
Perhaps there are some who do understand this, but I think UVtools is for all of us, correct? Or just for a bunch of chosen ones?
Does everyone need to know how to write a computer program...?

So re-assuming.

If I want to grade the area I type 10 mm2 as minimum, and 1000 mm2 as maximum, correct?
That seems logical.

'proportional' refers to 'dependant on the size'? am I correct?

Bottom Layers:

- Bottom height is obvious: up to 0.25mm the layers will be considered as bottom ones.

- Limits: what is this? different layer thickness for bottom layers in one CTB? Because that's what logic tells in this case.
Of course this does not make sens, so what are these values for?

- Limits layers: 0.25 mm of the base height, and we can set 1 to 100 layers? 
Does it mean that the layer height will be adjusted to the thickness divided by nr of layers?
How the program determines which value to choose? Minimum or maximum?
What is the point of this?

- Apply when: which option is for simple 'apply with no questions asked'?

**Reply by sn4k3** [reply]
> Yes, you need to work on being clear with what you are offering in the program.
Having control over the variety of areas in layers is logical, just the explanation is not understandable for everyone.
Perhaps there are some who do understand this, but I think UVtools is for all of us, correct? Or just for a bunch of chosen ones?
Does everyone need to know how to write a computer program...?

A program that is available for everyone doesn't make it for everyone. Simply there are people who never get how to use it or don't fit with it.
I can give my own example: Photoshop, for me one of the most complicated programs I ever used, I can't even do basic stuff without see one tutorial, beside I have the program for more than a decade I never able to use it nor invest time in learning it. This to explain that programs are not for everyone even if everyone can use it.

> If I want to graduate the area I type 10 mm2 as minimum, and 1000 mm2 as maximum, correct?

No, there are no limits for area, only time (s)
For area you need to set Bottom area and Normal area. Look at top **Bold** Labels

> Bottom height is obvious: up to 0.25mm the layers will be considered as bottom ones.

If you read the Jan article in detail the study shows that the closer you are from LCD the more it resists making it harder to settle on desired Z. Nothing prevents you from use 0.25mm but that is too close to be a "good value" unless you have very thin resin and good machine with strong Z axis. 1mm is a safe value with a margin for worst machines.

> Limits: what is this? different layer thickness for bottom layers in one CTB? Because that's what logic tells in this case.
Of course this does not make sens, so what are these values for?

Explained in first reply. They are in seconds not mm

> Limits layers: 0.25 mm of the base height, and we can set 1 to 100 layers?
Does it mean that the layer height will be adjusted to the thickness divided by nr of layers?
How the program determines which value to choose? Minimum or maximum?
What is the point of this?

If you refer to Base Height setting you can only set it in millimeters. Makes no sense to select a layer count as that give variable height depending on layer height. The explanation is the same as above: "...Jan article in detail the study shows that the closer you are from LCD the more it resists making it harder..."
Number of base layers will be given by: `Layer Height / Base Height`. That means at 0.05mm will result in 20 layers but at 0.10mm will give 10 layers. But in every case the bottom wait time is applying up to 1mm no matter the layer thickness -- This is the most correct.

> Apply when: which option is for simple 'apply with no questions asked'?

"Auto apply" check will auto apply suggestion on file load without any question.

**Comment by Hazoooka** [comment]
- The menu is not saying anything about seconds as you can see in the screenshot.
![Screenshot 2023-06-21 at 23 09 32](https://github.com/sn4k3/UVtools/assets/137073500/40d5bab7-dab7-4454-bdf2-17d8a1aac449)

- where can I find 'auto apply'? there is no button or selection in the menu
![Screenshot 2023-06-21 at 23 12 17](https://github.com/sn4k3/UVtools/assets/137073500/b76b1832-60f8-49fb-a719-08928b70da2b)

**Reply by sn4k3** [reply]
You don't clear mention you talking about the "Bottom Layers" **suggestion**, you started the talk by "proportional" and "mm^2" which are from Wait time suggestions. As Wait times also have settings for bottom layers, your "Bottom Layers:" indication is insuficiente to make me think you swap to another suggestion. Always use the full terms in order to avoid confusion.

That suggestion "enforces" a number of bottom layers you should have on file.
You define bottom layers in mm, where it will calculate the number of layers by: `Bottom height / Layer Height`, also it's constrained by the first limit "Limits (mm)"
Then the resulting layer count will be constrained with the "Limits (Layers)"

> where can I find 'auto apply'? there is no button or selection in the menu

![image](https://github.com/sn4k3/UVtools/assets/113281/105f7c1a-2df4-42df-bc8e-c0ff98ed01eb)

That will auto apply everytime even if you manually change bottom layer count, the suggestion will kick-in and replace the value again without any prompt

**Comment by Hazoooka** [comment]
> If you refer to Base Height setting you can only set it in millimeters. Makes no sense to select a layer count as that give variable height depending on layer height. The explanation is the same as above: "...Jan article in detail the study shows that the closer you are from LCD the more it resists making it harder..."
> Number of base layers will be given by: `Layer Height / Base Height`. That means at 0.05mm will result in 20 layers but at 0.10mm will give 10 layers. But in every case the bottom wait time is applying up to 1mm no matter the layer thickness -- This is the most correct

> A program that is available for everyone doesn't make it for everyone. Simply there are people who never get how to use it or don't fit with it.

Of course. It is your duty to explain people how to use it. In the manual for example.

> No, there are no limits for area, only time (s)
> For area you need to set Bottom area and Normal area. Look at top **Bold** Labels

Yes, correct, bottom and normal.
No limits for the area, and yet you can select the size. What for? for the time?
The time is set in the other two lines.

So, Wait time before print:
- what does wait time (fallback) mean?
- what does 'proportional wait time' mean? why 'proportional' and not just time?
- why 'proportional area' and not just area? what 'proportional' mean?
- proportional > to previous mass -  do we leave it unchanged in usual situations? how those options affect the print?

**Reply by sn4k3** [reply]
> Of course. It is your duty to explain people how to use it. In the manual for example.

No, it's not, as I already told I make the program for hobby and fun. I have 0 obligations here and also you use the program at your own risk. If it was a paid product, you may have that reason, but it's not the case. I don't even care if people use it or not. I will keep build it as long I have fun.
If I'm still chatting with you is not because I must but is because I want to.

> Yes, correct, bottom and normal.
No limits for the area, and yet you can select the size. What for? for the time?
The time is set in the other two lines.

Area don't need limits because in theory can be "infinite" and must relate with time. The wait time is what is set to file and so it makes only sense to have a limit in what you are setting. All values are calculated with infinite bounds, only in the end the value that will be set is limited. 

> what does wait time (fallback) mean?

Fallback is fixed time when is not possible to apply wait time per layers. Many file formats are not able to set time per layer, which will use that fixed and global value (All bottoms and normals will have same time).

> what does 'proportional mean wait time' mean? why 'proportional' and not just time?
> why 'proportional area' and not just area? what 'proportional' mean?

Explained on first reply, go back and read again. But in resume it's because that field value will be proportional to another field value, which is the time - mass.

> proportional > to previous mass - do we leave it unchanged in usual situations? how those options affect the print?

Previous mass is the best setting IMO, It means it will get the mass of the previous layer to calculate the time to set on actual layer. As the current layer need to settle before print, the mass of previous layer is what influence the EF. For deeper tanks full of resin maybe the other options are better.

> limits (bottom)/(normal): is it also in seconds? Or layer count?

Seconds, look at the grey unit on the right. Every field have a unit on the right to indicate what they are.

![image](https://github.com/sn4k3/UVtools/assets/113281/af35d51f-ae3e-4728-a5ff-4ede3088b2a8)

**Comment by Hazoooka** [comment]
> I have 0 obligations here and also you use the program at your own risk.
> I don't even care if people use it or not. I will keep build it as long I have fun.

By making your program public, getting so many users, and cooperating closely with Lychee Slicer team you take full responsibility. It goes automatically. You have full obligation to the community.
You do not care about people using your program, interesting.
My advice, you better tell it to Lychee and users straight because clearly they can not rely on you. Be fair.
But it's your problem to solve.

> Seconds, look at the grey unit on the right. Every field have a unit on the right to indicate what they are.

Correct, missed this one.

Transition layers:
- transition step: in seconds, it's clear
- limit layers: I presume it's in seconds? or layer count? because the description does not say! However you mention the exposure time so it probably in seconds 
![image](https://github.com/sn4k3/UVtools/assets/113281/693af6bf-d613-42a1-8488-352b60a5ac61)

**Reply by sn4k3** [reply]
> By making your program public, getting so many users, and cooperating closely with Lychee Slicer team you take full responsibility. It goes automatically. You have full obligation to the community.

Wrong. I do not imply the software to anyone, you use it because you want to, UVtools is 100% not required to print and none product depends on it. I'm not selling a product or association. 
While I chat with lychee we don't have any agreement, both are independent from each other and free to decide on their own, UVtools is a free and open-source program, which you can decide without my approval to call in your application or not if inside the scope of the license. You can also grab my work, modify it at your will, do a better or worse program and distribute it (If license is followed), that's the nature of this project. 

> You do not care about people using your program, interesting.

Correct, it's a hobby It don't rely on others to keep going and built without any interest other than fun. 
So, if you and 99% more stop to use UVtools I don't care at all. If a better solution come to replace UVtools (Slicer that does what UVtools do inbuilt), I don't care, this is my personal project and you have to understand that I do what I want with it, without have to answer for anything or to anyone. This is about me and having fun with programming.

You may not know, but I only have 1 printer that I rarely use, I really dislike resin printing due the post-process and chemicals, this something that isn't for me. And once again why I build UVtools? For fun and programming challenges. Yet you have one masterpiece of software. 
Mostly when you do things for money or second interests all fun will be gone, so I like to keep it simple, free and in the terms I want. 
If you do not agree with the software policy then stop using it. Either way I don't care...

> clearly they can not rely on you

They can, without any obligation or expectation. I'm free to answer as I'm also free to not answer.
Support-wise and the free nature of the program I can tell you, even paying on most of other products you will get much worse support than you getting out here and for free! The thing is I enjoy to talk about the software, again one of my pleasures -- This is the only motive why I'm still communicating with you, because let's be honest neither I or You care about each other and that's ok.
Tech-wise I'm biased but I can also tell you, for what it does one of the most reliable and optimized softwares out there. You can call it ugly but never bad or mis-performant.

> limit layers: I presume it's in seconds? or layer count? because the description does not say! However you mention the exposure time so it probably in seconds

Limits (Layers), are in layer count. The 3 vertical stacked dashes are the icon for layers but tooltip already tell that...

![image](https://github.com/sn4k3/UVtools/assets/113281/6df29585-7642-4921-9742-5493428bb1ce)

**Comment by Hazoooka** [comment]
Clearly you do not have an idea of what's private and when it becomes a public service.
You need to learn a lot in you life, but I have a feeling you have a lot of decades ahead of you, so you've got time.
Just remember that as long as you do it for decent reasons, you help people, it's all OK.
But If you are just a ballerina seeking the attention and admiration, pretending to be so chill and detached from others, you may need to consider getting help.

Thank you for doing your job for now.

Kind regards,
Witold.

**Reply by sn4k3** [reply]
Once again, I disagree, as creator of the product I'm impressed how you know more than myself what the terms are. 
Clearly you are not aware of the terms in what this project was made, worse is that you think I'm now always in dept and own users anything. You speak like you are my boss ("My Job this, my job that, it's your job, etc"), sad news - You aren't! You use this program for your own benefit and favor, as I see it here, you are the only who owns, not me... 

You need to understand when you create something, no matter what is, it comes in the terms you want.
Let's ask the neutral AI:
![image](https://github.com/sn4k3/UVtools/assets/113281/41d53399-61ac-4f16-95f7-12e42fbbec21)
`1) Is what I currently do.` but still I'm not obligated to anything.

TLDR: You use the software by your own risk without any guarantee of support.

All that said I raise my right and cease to provide you any future support (Unless you smart to change your nick and the way to talk) as clearly you do not respect the terms and the scope of the project.

**Comment by Hazoooka** [comment]
Of course you cease the support for me, I expect no less of you.
But thank you for what you have done by now.

Keep up a good work out there.

Kind regards,
Witold.

**Comment by Hazoooka** [comment]
Oh, before I "cease" contacting you ;)

Regulations are for people not knowing right from wrong, or people looking for an excuse to bail out.

Before you sit back comfortably about not being obliged to do anything for the users, and happy about not caring about them, let me remind you that your UVtools is officially hard coded as an option in a commercial application called Lychee Slicer by Mango3D.
This is not your private project anymore.
Just for the records, to give you an update.

**Reply by sn4k3** [reply]
A [few posts](https://github.com/sn4k3/UVtools/discussions/733#discussioncomment-6246775) behind have written about that too. And again, you wrong... You keep talking and talking about stuff you have not a clue about.
First you have to read and know [AGPL licence](https://github.com/sn4k3/UVtools/blob/master/LICENSE) which is what UVtools uses as it open-source license.

UVtools is not hard coded into lychee, i doubt if you even know what "hard coded" into means, but anyway, they give the option to open the resulting sliced file into UVtools which user MUST install by his own. Lychee does not distribute, include or force UVtools to user.
License allows any application, commercial or not to do what they want with my software (even copy it and modify it), AS LONG: License is followed. Which lychee and other software like tango does. Even you can use UVtools in your application without any of my approval, again if licence is followed.

In the case you don't to read license, let me resume for you:

![image](https://github.com/sn4k3/UVtools/assets/113281/db44dff7-eaae-4d21-8fd0-34aab0775951)

**You will have permissions to (Without any of my consent)**
- Commercial use
- Modification
- Distribution
- Patent use
- Private use

**By using it you and your users will NOT get:**
- Liability
- Warranty

Is now that clear? Can't be as transparent as that, LICENSE is self-explanatory and mandatory on the terms. No rocket science here...

> This is not your private project anymore.

To clear this out, the project is public domain, again license... But I still have the rights under the project. For example, I can't turn past versions privates as they are released under AGPL, but at any time as the owner that grant me the power to release future versions under another license and kept it private or even monetize it. As long I don't transfer the ownership this is still my project ;)

**Comment by Hazoooka** [comment]
OK, you really do not understand the case, you only understand licenses, regulations, fake… So be it.
It’s your life, not mine.

End of the case.

Stay safe!
Kind regards,
Witold.

**Reply by sn4k3** [reply]
> It’s your life, not mine.

First thing I agree with you :D
Peace!

**Comment by RabbitRoom** [comment]
Read your discussion very carefully. Thank you for all you do for us.
But, I would like to add a little bit.
In all sections there is a choice:
Outside limits: Applies when the recommended value does not meet the current limit criteria.
It is missing the option to apply the recommended values when the current values in the file do not meet the current limit criteria.
This would be very useful.

**Comment by Hazoooka** [comment]
Actually there is.
It's this greyed out 'Save' button.
In the configuration like on the screenshot you just need to click it after you're done with the changes.

I have missed this the first time, that's why the whole fuzz was about.
Well, nobody's perfect ;)
![Screenshot 2023-11-23 at 15 22 31](https://github.com/sn4k3/UVtools/assets/137073500/d67326e3-7d65-4a7a-8149-9e92533dfabe)
